src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -240,6 +239,9 @@ void PipelineExecutor::executeStepImpl(size_t thread_num, std::atomic_bool * yie
                 tasks.pushTasks(queue, async_queue, context);
             }
 
+            // Upscale if possible
+            spawnThreads();
```

Comment:
Ok!

Reply:
After implementing lock-free tryAcquire it looks unnecessary and also complicated because `num_threads` is not available in this function (requires either a new member `num_threads` or changing signature of couple private functions)